client/src/components/addRecipe/AddRecipeForm.js

```diff
@@ -1,33 +1,47 @@
-import { Form, InputGroup, Row, Button } from "react-bootstrap";
-import { Formik } from "formik";
-import * as yup from "yup";
+import { Form, InputGroup, Row, Button, Container, Col } from "react-bootstrap";
+import { Field, Formik } from "formik";
 import Input from "../UI/Input";
-import axios from "axios";
-import { useState } from "react";
+import { useContext, useState } from "react";
 import { Prompt, useHistory } from "react-router";
-
-import { BsFillPlusCircleFill } from "react-icons/bs";
-
+import { FaPlusCircle } from "react-icons/fa";
 import { inputs, inputsIngr } from "../../constants";
 import DNDlist from "./DNDlist";
+import AuthContext from "../../store/auth-context";
+import Title from "../UI/Title";
+import SubmitButton from "../UI/SubmitButton";
+import schema from "../../validations/recipeSchema";
+import useFetch from "../../hooks/useFetch";
 
 const baseUrl = process.env.REACT_APP_BASE_URL;
 
 const AddRecipeForm = () => {
   const history = useHistory();
+  const authCtx = useContext(AuthContext);
+
+  const url = `recipes/categories`;
+
+  const options = {
+    method: "GET",
+  };
+  const { data } = useFetch(url, options);
+  const categories = data?.categories;
 
   const [isFocusedForm, setIsFocusedForm] = useState(false);
 
   const [ingredients, updateIngredients] = useState([]);
   const [instructions, updateInstructions] = useState([]);
 
+  const [isLoading, setIsLoading] = useState(false);
+  const [error, setError] = useState([]);
```

Comment:
you could check yourself and add a custom hook useAsyncAction that you send the action and it knows to setLoading to true and error to null and update them as you did for every action but at one place

Reply:
u saw my useFetch? I got a little confused because I couldn't use useFetch inside a function (like addRecipe, login, register) so I'll read about "useAsyncAction"